src/coreComponents/linearAlgebra/DofManager.cpp

```diff
@@ -140,59 +138,119 @@ DofManager::Location DofManager::location( string const & fieldName ) const
+      fieldNames[MeshHelper< LOC >::syncObjName].emplace_back( field.key );
 
-    // synchronize across ranks
-    std::map< string, string_array > fieldNames;
-    fieldNames[MeshHelper< LOC >::syncObjName].emplace_back( field.key );
-    DomainPartition & domain = dynamicCast< DomainPartition & >( m_mesh->getParent().getParent().getParent() );
-    CommunicationTools::getInstance().synchronizeFields( fieldNames, *m_mesh, domain.getNeighbors(), false );
+      // register index array
+      helper::template create<>( mesh, field.key, field.docstring, regions );
+      typename helper::Accessor indexArray = helper::get( mesh, field.key );
+
+      // populate index array using a sequential counter
+      forMeshLocation< LOC, false, serialPolicy >( mesh, regions, [&]( auto const locIdx )
+      {
+        helper::reference( indexArray, locIdx ) = field.rankOffset + field.numComponents * index++;
+      } );
+
+      // synchronize across ranks
+      CommunicationTools::getInstance().synchronizeFields( fieldNames, mesh, m_domain->getNeighbors(), false );
+    } );
   } );
 }
 
 void DofManager::removeIndexArray( FieldDescription const & field )
 {
   LocationSwitch( field.location, [&]( auto const loc )
   {
-    Location constexpr LOC = decltype(loc)::value;
-    ArrayHelper< globalIndex, LOC >::template remove<>( *m_mesh, field.key, field.regions );
+    forMeshSupport( field.support, *m_domain, [&]( MeshBody &, MeshLevel & mesh, auto const & regions )
+    {
+      Location constexpr LOC = decltype(loc)::value;
+      ArrayHelper< globalIndex, LOC >::template remove<>( mesh, field.key, regions );
+    } );
   } );
 }
 
+void DofManager::computeFieldDimensions( localIndex const fieldIndex )
+{
+  FieldDescription & field = m_fields[fieldIndex];
+
+  // determine number of local support points
+  localIndex numLocalSupport = 0;
+  forMeshSupport( field.support, *m_domain, [&]( MeshBody &, MeshLevel & mesh, auto const & regions )
+  {
+    numLocalSupport += countMeshObjects< false >( field.location, mesh, regions );
+  } );
+  field.numLocalDof = field.numComponents * numLocalSupport;
+
+  // gather dof counts across ranks
+  field.rankOffset = MpiWrapper::prefixSum< globalIndex >( field.numLocalDof );
+  field.numGlobalDof = field.rankOffset + field.numLocalDof;
+  MpiWrapper::broadcast( field.numGlobalDof, MpiWrapper::commSize() - 1 );
+
+  // determine field's offsets
+  if( fieldIndex > 0 )
+  {
+    FieldDescription const & prev = m_fields[fieldIndex - 1];
+    field.blockOffset = prev.blockOffset + prev.numGlobalDof;
+  }
+  field.globalOffset = field.rankOffset; // actual value computed in reorderByRank()
+}
+
 namespace
 {
 
-std::vector< string > processFieldRegionList( MeshLevel const & mesh,
-                                              std::vector< string > const & inputList )
+std::vector< string >
+processFieldRegionList( MeshLevel const & mesh,
+                        std::vector< string > inputList )
```

Comment:
is `inputList` a `regionNames`?

Reply:
Indeed, it's a list of region names. If not empty, the function checks that every name corresponds to an existing region and returns it unchanged. If empty, it builds and returns a list of all region names, which allows the user to skip listing all regions when there's many of them.

Reply:
OK; then I presume `inputList` should be renamed into `regionNames` 😉